src/tsdb.c

```diff
@@ -201,23 +220,109 @@ size_t SeriesGetNumSamples(const Series *series) {
     return numSamples;
 }
 
+static void upsertHandleRules(Series *series, AddCtx *aCtx) {
+    CompactionRule *rule = series->rules;
+    while (rule != NULL) {
+        // upsert in latest timebucket
+        if (aCtx->sample.timestamp >= CalcWindowStart(series->lastTimestamp, rule->timeBucket)) {
+            if (aCtx->sz < 1) { // a sample was updated or removed
+                rule->backfilled = true;
+            }
+            continue;
+        }
+        rule->backfilled = true; // TODO: fix so handleCompaction does not append but still use on-the-fly agg context 
+        timestamp_t start = CalcWindowStart(aCtx->sample.timestamp, rule->timeBucket);
+        // ensure last include/exclude
+        double val = SeriesCalcRange(series, start, start + rule->timeBucket - 1, rule->aggClass);
+        // if (isnan(val)) return TSDB_ERROR;
+
+        RedisModuleKey *key;
+        Series *destSeries;
+        if (!GetSeries(RTS_GlobalRedisCtx, rule->destKey, &key, &destSeries, REDISMODULE_READ)) {
+            // TODO: log something
+            continue;
+        }
+        SeriesAddSample(destSeries, start, val);
+        RedisModule_CloseKey(key);
+        rule = rule->nextRule;
+    }
+}
+
+static int SeriesUpsertSample(Series *series, api_timestamp_t timestamp, double value) {
+    void *chunkKey = NULL;
+    Chunk_t *chunk = series->lastChunk;
+    timestamp_t lastChunkFirstTS = series->funcs->GetFirstTimestamp(series->lastChunk);
+
+    if (timestamp < lastChunkFirstTS) {
+        // Upsert in an older chunk
+        timestamp_t rax_key;
+        seriesEncodeTimestamp(&rax_key, timestamp);
+        RedisModuleDictIter *dictIter = 
+            RedisModule_DictIteratorStartC(series->chunks, "<=", &rax_key, sizeof(rax_key));
+        chunkKey = RedisModule_DictNextC(dictIter, NULL, (void*)&chunk);
+        if (chunkKey == NULL) {   // should not happen since we always have a chunk
+            // TODO: check if new sample before first sample
+            RedisModule_DictIteratorStop(dictIter);
+            assert(0);
+            return TSDB_ERROR;
+        }
+    }
+    Sample sample = {.timestamp = timestamp, .value = value };
+    AddCtx aCtx = { .sz = 0,
```

Comment:
`AddCtx` is a problematic pattern, its both arguments and return value(s?) from Upsert. 
please split it to avoid any confusion between them.

Reply:
Will expand explanation on the struct.